### PR TITLE
Fix sharing same authentication token between sender and recipient

### PR DESCRIPTION
**Pull request by sameera207**
**Issue**
---
Previously when a secret is created , only one authenticated token is created. So when the secret is viewed by the recipient , the same authentication token is  used to validate the user. This leads to identify  recipient as the sender for future secret creations. 

**Fix**
---
Now when creating a new secret, (for the first time), 2 authentication tokens are creating for both sender and recipient and authentication is done through them. 

**New workflow**
---
**Creating the secret for the first time**

_when entering both sender and receiver_

- Auth token is created for sender
- Auth token is created for receiver
- Empty secret is created

_when entering only the sender_

- Auth token is created for sender

here on, both users will be authorized by their own tokens

**NOTE**

4 request specs are currently failing due to `js` validations are failing. I've checked them as a normal user and they all work in the app itself. its just that , `capybara + rspec + js` issue. I'll have a look whats going with them.

### CODE REVIEW

File: app/services/auth_token_service.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    auth_token = AuthToken.new(auth_hash).generate
+
+    if recipient_email
+      # generate a seperate authentication for the recipient
+      AuthToken.new(email:  recipient_email).generate
+
+      # cretae an empty secret
+      access_key = SecureRandom.hex(10)
+      SecretService.create_empty_secret(auth_hash['email'],
```
Comment (ralovely) [inline app/services/auth_token_service.rb]: Creating a Secret in the authToken service/controller seems like a tight coupling and a mix of responsibility.
At minimum, there is a naming problem: `AuthTokenService.generate` doesn't suggest in any way that I will also end up with a secret.
The check for recipient_email and the creation of the secret should maybe move to the controller, making it more obvious.

Better, I think the creation of the Secret should generate the Token for the recipient, if needed (which will always be the case, but just happening at different moment depending on the path)

**Comment by sameera207** [comment]
@ralovely , for this Friday PD ... :smile:

File: app/services/secret_service.rb [hunk 2]
```diff
@@ -2,12 +2,30 @@ class SecretService
 
   # The Secret#secret_key value isn't persisted - it is used to encrypt the secret
   # and the only copy of it is sent in the notification email to the recipient.
-  def self.encrypt_secret(params, request_host)
+  def self.encrypt_existing_secret(id, params, request_host)
```
Comment (ralovely) [inline app/services/secret_service.rb]: Why aren't all these (`encrypt_existing_secret`, `encrypt_new_secret` and `create_empty_secret`) just methods on the Secret class ?

and the 2 encrypt method seem very redundant.

```
s = Secret.new(params...)
s.encrypt
```
seems a more natural construct.

File: app/views/secrets/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<div class="col-sm-12">
+  <div class="text-center">
+    <h2>Create a new secret to send</h2>
+  </div>
+
+  <div class='row'>
+    <div class="col-md-6">
+      <%= simple_form_for @secret, url: secret_url, wrapper: 'horizontal_form', html: {class: 'form-horizontal'} do |f| %>
```
Comment (ralovely) [inline app/views/secrets/edit.html.erb]: There is probably a way to put this whole form (most of the page really) in a partial and DRY it up with the `#new` view

**Pull request merged**